### PR TITLE
Task API improvements, task engine processing explained.

### PR DESCRIPTION
**Pull request by tomaszmichalak**
## Description
Task API and Task Engine documentation updates. This PR explains how single and composite nodes are processed and improves Task API docs.

## Motivation and Context
Improve technical documentation.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/Knotx/knotx/blob/master/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: task/api/README.md [hunk 1]
```diff
@@ -1,34 +1,43 @@
...
+- task declaration syntax, check the [Task Factories](https://github.com/Knotx/knotx-fragments/tree/master/task/factory)
+- task processing details (such as error handling etc.), check the [Task Engine](https://github.com/Knotx/knotx-fragments/tree/master/task/engine)
 
 ## Task
-Task decomposes business logic into lightweight dependent parts. Those parts are graph [nodes](#node) 
-connected by [transitions](#transition). So a task is a directed graph of nodes. Nodes define an operation on the 
-fragment (some business logic), whereas transitions draw business decisions.
-This graph is acyclic and each of its nodes can be reached only from exactly one path (transition). 
-These properties enable us to treat the Task as a tree structure.
+Task transforms a [fragment](https://github.com/Knotx/knotx-fragments/tree/master/api#fragment) to
```
Comment (Voycawojka) [inline task/api/README.md:9]:
```suggestion
Task transforms a [fragment](https://github.com/Knotx/knotx-fragments/tree/master/api#fragment) into 
```

File: task/api/README.md [hunk 2]
```diff
@@ -1,34 +1,43 @@
...
-connected by [transitions](#transition). So a task is a directed graph of nodes. Nodes define an operation on the 
-fragment (some business logic), whereas transitions draw business decisions.
-This graph is acyclic and each of its nodes can be reached only from exactly one path (transition). 
-These properties enable us to treat the Task as a tree structure.
+Task transforms a [fragment](https://github.com/Knotx/knotx-fragments/tree/master/api#fragment) to 
+a new one. It declares [fragment operations](https://github.com/Knotx/knotx-fragments/tree/master/api#fragment-operation) 
+to be performed and connections between them. It reflects that a task is a graph, whose [nodes](#node) 
+are fragment operations linked by edges (called [transitions](#transition)). So a task is a directed 
+graph of nodes. Besides, each node can be reached only from exactly one path (transition), so the graph 
+is acyclic, basically it is a tree.
```
Comment (Voycawojka) [inline task/api/README.md:14]:
```suggestion
graph of nodes. Besides, each node can be reached only from exactly one path (transition), meaning the graph 
is acyclic. It's basically a tree.
```

File: task/api/README.md [hunk 3]
```diff
@@ -1,34 +1,43 @@
...
+Tasks allow defining more [complex structures](#composite-node) such as a list of subtasks (subgraphs).
 
 ### Node
-Node has a single input and declares zero-to-many outputs (transitions). Node outputs (transitions) 
-indicate business decisions, whereas node logic can modify a fragment.
+Node has a single input and declares zero-to-many outputs (transitions). There are two base node 
+types: a [single node](#single-node) and a [composite node](#composite-node).
+
+See also the [Actions Library](https://github.com/Knotx/knotx-fragments/tree/master/action/library), 
+that provides the most common node's implementations e.g `HTTP Action`.
```
Comment (Voycawojka) [inline task/api/README.md:28]:
```suggestion
that provides the most common node's implementations, e.g `HTTP Action`.
```

File: task/api/README.md [hunk 4]
```diff
@@ -37,47 +46,67 @@ found or even expired password.
 
 #### Composite Node
 Composite node defines a list of subgraphs to evaluate. It enables `parallel` processing of independent 
-nodes/subgraphs (e.g. calling two external independent data sources).
+subgraphs (e.g. calling two external independent data sources). Each subgraph gets a fragment copy and 
+updates a fragment's payload. When all subgraphs end, then the reduction phase happens during which 
+all subgraphs results are merged (all copies' payloads).
```
Comment (Voycawojka) [inline task/api/README.md:51]: So are we calling them subgraphs or subtasks? The word `subtask` was already introduced and now it's not used

File: task/api/README.md [hunk 5]
```diff
@@ -37,47 +46,67 @@ found or even expired password.
...
+  The diagram above represents a composite node that consists of: 
+  - subgraph starting with `A` node
+  - subgraph that is also a composite node (that consists of two subgraphs starting with `C` and `D` nodes).
 
-or other composite nodes (or a mix of both).
+so it is very elastic.
 
-![Nested composite nodes](assets/images/nested-composite.png)
+Despite the fact, that composite node is more like subgraph structure its general purpose is to 
+transform one fragment into a new one. Thinking more abstract, it follows the same rules as
```
Comment (Voycawojka) [inline task/api/README.md:74]:
```suggestion
transform one fragment into a new one. Thinking more abstractly, it follows the same rules as 
```

File: task/engine/README.md [hunk 6]
```diff
@@ -1,78 +1,106 @@
...
-## Fragment's status
-During fragment's processing, a fragment's status is calculated. Each node responds with a transition. 
-Fragments Engine validates node responses and set one of the fragment's statuses:
-- `unprocessed`
+## How does it work
+Fragments Engine accepts a list of fragments decorated with the incoming HTTP request. Then
+it converts the list to [Reactive Stream](https://github.com/reactive-streams/reactive-streams-jvm) 
+and processes all its item independently, in parallel (`map`). Each time a fragment processing ends, 
+the result is added to the final result. Finally, it returns a list of processed fragments with the 
+initial order.
```
Comment (Voycawojka) [inline task/engine/README.md:13]:
```suggestion
and processes all its items independently, in parallel (`map`). Each time a fragment processing ends, 
its' result is added to the final result. Finally, the engine returns a list of processed fragments with the 
initial order.
```

File: task/api/README.md (hunk 4 above)
Comment (tomaszmichalak) [reply task/api/README.md:51]: Replaced all `subgraphs` with `subtasks`.

**Pull request merged**